### DISCUSSION TITLE
[help] targets equivalent of dplyr split-apply-combine workflow

### DISCUSSION THREAD

**Discussion by uhkeller** [ropensci/targets#1082] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm trying to switch from a tidyverse-based split-apply-combine workflow to targets and I'm struggling to understand how to do this properly. Since the operations I'm doing on each of the (dplyr) groups take quite a long time I want to use dynamic branching to get the benefits of caching and parallel execution.

In dplyr I would do something like:

``` r
library(dplyr, warn.conflicts = FALSE)
iris %>% 
  group_by(Species) %>%
  summarise(m = mean(Sepal.Length))
#> # A tibble: 3 × 2
#>   Species        m
#>   <fct>      <dbl>
#> 1 setosa      5.01
#> 2 versicolor  5.94
#> 3 virginica   6.59
```

The most straight forward approach in targets seemed to be this:

``` r
library(targets)
tar_script({
  library(dplyr, warn.conflicts = FALSE)
  list(
    tar_target(
      data,
      iris %>% group_by(Species) %>% tar_group(),
      iteration = "group"
    ),
    tar_target(
      summary,
      data %>% summarise(m = mean(Sepal.Length)),
      pattern = map(data),
      iteration = "group"
    )
  )
}, ask = FALSE)
tar_make(reporter = "silent")
tar_read(summary)
#> # A tibble: 3 × 1
#>       m
#>   <dbl>
#> 1  5.01
#> 2  5.94
#> 3  6.59
```

But now there is no easy way to tell which mean belongs to which level of `Species`. This is because `tar_group()` explicitly drops the grouping. One simple workaround is to define my own version of `tar_group()` that does not drop the dplyr grouping, which works in this case — but it might not in others as I assume there is a reason for the grouping being dropped.

``` r
library(targets)
tar_script({
  my_tar_group <- 
    function (x) 
    {
      tar_assert_package("dplyr")
      groups <- attr(x, "groups")
      if (is.null(groups)) {
        tar_throw_validate("tar_group() must take a grouped data frame from dplyr::group_by()")
      }
      x$tar_group <- targets:::tar_group_column(x, groups)
      # Commented out the lines that drop the dplyr grouping
      # attr(x, "groups") <- NULL
      # class(x) <- setdiff(class(x), "grouped_df")
      x
    }
  library(dplyr, warn.conflicts = FALSE)
  list(
    tar_target(
      data,
      iris %>% group_by(Species) %>% my_tar_group(),
      iteration = "group"
    ),
    tar_target(
      summary,
      data %>% summarise(m = mean(Sepal.Length)),
      pattern = map(data),
      iteration = "group"
    )
  )
}, ask = FALSE)
tar_make(reporter = "silent")
tar_read(summary)
#> # A tibble: 3 × 2
#>   Species        m
#>   <fct>      <dbl>
#> 1 setosa      5.01
#> 2 versicolor  5.94
#> 3 virginica   6.59
```

Any suggestions?

**Comment by wlandau** [comment]
`tar_group()` drops the `dplyr` grouping attributes because they are dropped anyway in efficient/compressed storage formats like `"fst_tbl"`, and I do not want this to come as a surprise for folks who change the `format` argument of `tar_target()`.

Rather than defining your own `tar_group()` preserve the attributes, you can explicitly assign more variables in the `summarise()` call:

```r
tar_target(
  summary,
  data %>% summarise(
    m = mean(Sepal.Length),
    Species = Species[1]
  ),
  pattern = map(data),
  iteration = "group"
)
```

**Reply by uhkeller** [reply]
Thanks, that makes sense. Since I'm using other features that `"fst_tbl"` does not support (list columns), this is not really an issue for me.

Adding the grouping variables to the `summarise()` call seems a bit tedious as I have to remember to add all the grouping variables (in the actual application there will be more than one), and also error-prone especially if the grouping changes at some point.

Since all that `tar_group()` really does seems to be adding the `tar_group` variable, I guess I'll just do this myself in a `mutate()` call.

``` r
library(targets)
tar_script({
  library(dplyr, warn.conflicts = FALSE)
  list(
    tar_target(
      data,
      iris %>% group_by(Species) %>% mutate(tar_group = cur_group_id()),
      iteration = "group"
    ),
    tar_target(
      summary,
      data %>% summarise(m = mean(Sepal.Length)),
      pattern = map(data),
      iteration = "group"
    )
  )
}, ask = FALSE)
tar_make(reporter = "silent")
tar_read(summary)
#> # A tibble: 3 × 2
#>   Species        m
#>   <fct>      <dbl>
#> 1 setosa      5.01
#> 2 versicolor  5.94
#> 3 virginica   6.59
```

Is there any reason not to do this?

Btw, I very much appreciate how incredibly helpful you are. It's amazing to see the issues for this repo (453 closed, 1 open)… Thank you very much for this, and also for `{targets}` and your other packages of course!

**Reply by wlandau** [reply]
Thanks @uhkeller!

> Is there any reason not to do this?

`tar_group = dplyr::cur_group_id()` should definitely work in your case. Others can use it if they can be confident that the `format` in `tar_target()` will not drop attributes of data frames.

**Reply by uhkeller** [reply, marked as answer]
Great, thanks! Since the information is a bit spread out I will summarize it here and mark this reply as the answer.

`tar_group()` removes `{dplyr}` groups because all storage formats except `"rds"` and `"qs"` remove them anyway (silently, at least in the case of `"fst"` and `"fst_tbl"`). In case one is sure that one will only use storage formats that retain the groups, one can use `mutate(data, tar_group = dplyr::cur_group_id())` instead of `tar_group(data)` to add the `tar_group` variable, retaining the `{dplyr}` groups. In all other cases, one has to explicitly add the grouping variables to the result of the `summarise()` call, in my example by adding `Species = Species[1]` to the call.
